shapeways/oauth2_client.py

```diff
@@ -51,18 +79,57 @@ def authenticate(self, client_id, client_secret):
     def _validate_response(self, response):
         """
         Internal function - validate results
-        :rtype: list()
-        """
-        if response.status_code != 200:
-            raise RuntimeError("Call threw status {}".format(response.status_code))
+        :rtype dict()
+        """
+
+        # Default values - these will be modified as needed by the rate limiting error handling
+        rate_limit = RateLimitingHeaders(
+            rate_limit_type=SW_RATE_LIMIT,
+            is_rate_limited=False
+        )
+        headers = response.headers
+
+        # Handle HTTP errors
+        if response.status_code != HTTP_OK:
+            if response.status_code == HTTP_RATE_LIMITED:
+                # We're rate limited
+                rate_limit.is_rate_limited=True
+
+                if CF_RATE_LIMIT_RETRY in headers.keys():
+                    # Limited by CF - backoff for a number of seconds
+                    rate_limit.rate_limit_type=CF_RATE_LIMIT
+                    rate_limit.rate_limit_remaining=0
+                    rate_limit.rate_limit_retry_inseconds=headers[CF_RATE_LIMIT_RETRY]
+                else:
+                    # Shapeways Rate limiting - stupidly, we move the retryInSeconds entry from the response headers
+                    # to the body.  Dealing with this here.
+                    rate_limit.rate_limit_remaining = 0
+                    rate_limit.rate_limit_retry_inseconds = response.json()['rateLimit']['retryInSeconds']
+
+                return {CONTENT_SUCCESS: False, CONTENT_RATE_LIMIT: rate_limit.__dict__}
+            else:
+                # Generic error
+                content = response.json()
+                content[CONTENT_SUCCESS] = False
+                content[CONTENT_RATE_LIMIT] = rate_limit.__dict__
+                return content
```

Comment:
Tiny recommendation to slightly reduce code complexity would be to remove unnecessary `else`s:

(untested)
```
# Handle HTTP errors
if response.status_code != HTTP_OK:
    # Generic error
    content = response.json()
    content[CONTENT_SUCCESS] = False
    content[CONTENT_RATE_LIMIT] = rate_limit.__dict__

    if response.status_code == HTTP_RATE_LIMITED:
        # We're rate limited
        rate_limit.is_rate_limited=True

        # Shapeways Rate limiting - stupidly, we move the retryInSeconds entry from the response headers
        # to the body.  Dealing with this here.
        rate_limit.rate_limit_remaining = 0
        rate_limit.rate_limit_retry_inseconds = response.json()['rateLimit']['retryInSeconds']

        if CF_RATE_LIMIT_RETRY in headers.keys():
            # Limited by CF - backoff for a number of seconds
            rate_limit.rate_limit_type=CF_RATE_LIMIT
            rate_limit.rate_limit_retry_inseconds=headers[CF_RATE_LIMIT_RETRY]

        content = {CONTENT_SUCCESS: False, CONTENT_RATE_LIMIT: rate_limit.__dict__}
        
    return content
```

Reply:
Not required change => ✅

Reply:
So i had that originally - trouble is that the json payload only has the rateLimitRetry key if and only if the rate limiter has been achieved.  Therefore, this statement cannot be treated as always valid

```rate_limit.rate_limit_retry_inseconds = response.json()['rateLimit']['retryInSeconds']```

Hence, extra elses